backend/fms_core/services/sample_lineage.py

```diff
@@ -25,3 +26,82 @@ def create_sample_lineage(parent_sample, child_sample, process_measurement):
             errors.append(str(e))
 
     return (sample_lineage, errors, warnings)
+
+def create_sample_lineage_graph(mid_sample):
+    errors = []
+
+    if not mid_sample:
+        errors.append(f"Sample is required for sample lineage graph creation")
+
+    nodes = [mid_sample]
+    edges = []
+
+    if not errors:
+        def add_lineage(process_measurements):
+            queryset = process_measurements.select_related("process").prefetch_related("lineage")
+            queryset = queryset.annotate(child_sample=F("lineage__child"))
+            queryset = queryset.annotate(child_sample_name=F("lineage__child__name"))
+            queryset = queryset.annotate(source_sample_name=F("source_sample__name"))
+            return queryset
+
+        def query_set_iterator(f, l):
+            for x in l:
+                try:
+                    yield f(x)
+                except ObjectDoesNotExist:
+                    yield None
+
+        # fetch children
+        stack = [mid_sample]
+        visited = {mid_sample.id}
+        while stack:
+            child_sample = stack.pop()
+            new_process_measurements = [
+                p
+                for p in add_lineage(child_sample.process_measurement.all())
+            ]
+
+            sample_lineages = [
+                sl
+                for sl in query_set_iterator(lambda p: p.lineage.get(), new_process_measurements)
+            ]
+
+            new_child_samples = [
+                sl.child
+                for sl in sample_lineages
+                if sl and sl.child.id not in visited
+            ]
+
+            nodes.extend(new_child_samples)
+            edges.extend(new_process_measurements)
+
+            stack.extend(new_child_samples)
+            visited.update(s.id for s in new_child_samples)
+        
+        def find_process_measurement(parent_sample, child_sample):
+            process_measurements = list(add_lineage(parent_sample.process_measurement.all()))
+            for p in process_measurements:
```

Comment:
searching iteratively in your queryset is very likely to be slower than using the queryset filtering capacity. This function seem to be processMeasurement.objects.filter(source_sample=parent_sample, child_sample=child_sample) with the extra annotations and select_related...